### PR TITLE
feat: backtracking graph implementation

### PR DESCRIPTION
**Pull request by BonyBabu**

### CODE REVIEW

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+            return sourceSlice.getNodePosition(sourceNode);
+        } else if (graphEdge.getEdgeColor().equals(EntityColor.RED)) {
+            return findNextNodeIndexIfCurrentIsRed(sourceSlice);
+        } else {
+            return sourceSlice.getNodePosition(graphEdge.getDestinationNode());
+        }
+
+    }
+
+    public boolean isBacktrackingRequired(int currentSliceIndex, GraphEdge<T> graphEdge) {
```
Comment (stephencurran) [inline decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: Many of the methods are public but I'm wondering if this needs to be.  From what you showed me earlier, the client just needs to call 1 or 2 methods.

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    private int currentSourceSliceIndex = -1;
+    private GraphEdge<T> graphEdge = null;
+    private GraphEdge<T> lastSuccessfulEdge = null;
+
+    public BackTrackingGraph(List<Slice> slices) {
+        this.slices = slices;
+        this.edges = new HashMap<>();
+    }
+
+    private int findNextNodeIndexIfCurrentIsRed(Slice sourceSlice) {
```
Comment (stephencurran) [inline decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: Can you comment these methods to make clear what they do?

File: decoder/pom.xml [hunk 3]
```diff
@@ -98,5 +98,11 @@
             <scope>test</scope>
         </dependency>
 
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <version>3.2.4</version>
+        </dependency>
```
[the remark this answers is not in the document]
Comment (BonyBabu) [reply decoder/pom.xml]: fixed

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java (hunk 1 above)
Comment (BonyBabu) [reply decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: fixed

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java (hunk 2 above)
Comment (BonyBabu) [reply decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: comments are added

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+    private GraphEdge<T, L> graphEdge = null;
+    private GraphEdge<T, L> lastSuccessfulEdge = null;
+
+    public BackTrackingGraph(List<Slice<T, L>> slices) {
+        this.slices = slices;
+        this.edges = new HashMap<>();
+    }
+
+    /**
+     * if the user has marked last edge returned by the graph as red;
```
Comment (stephencurran) [inline decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: No need for `;` at the end of line

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+     * then the next edge should be an unmarked edge with the same start node
+     * But If all the edges starts from this start node is already marked
```
Comment (stephencurran) [inline decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: But **if** all the edges **starting** from this start node **are** already marked

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+     *
+     * @param sourceNode
+     * @param sourceSlice
+     */
+    private void handleIncomingEdges(GraphNode<T, L> sourceNode, Slice sourceSlice) {
+        if (sourceNode.isDeadEnd() && slices.indexOf(sourceSlice) != slices.size() - 1) {
+            sourceNode.getIncomingEdge().forEach(
+                    edge -> {
+                        edge.setEdgeColor(EntityColor.RED);
+                    }
```
Comment (stephencurran) [inline decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: No need for `{` and `}` in lambda.

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+            );
+        }
+    }
+
+    /**
+     * Backtracking is required if the green edges in the previous slices are not connected directly or indirectly to the given edge
+     *
+     * @param currentSliceIndex index of the slice from which the edge starts
+     * @param graphEdge         edge of which the predecessors need to verified
```
Comment (stephencurran) [inline decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: need to **be** verified

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+     */
+    public boolean isBacktrackingRequired(int currentSliceIndex, GraphEdge<T, L> graphEdge) {
+        int sliceIndex = currentSliceIndex;
+        for (GraphEdge<T, L> edge = graphEdge; edge != null; --sliceIndex) {
+            edge = edge.getSourceNode().getIncomingEdge()
+                    .stream()
+                    .filter(incomingEdge -> incomingEdge.getEdgeColor().equals(EntityColor.GREEN)).findFirst().orElse(null);
+        }
+        return (sliceIndex != -1);
+
```
Comment (stephencurran) [inline decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: empty line

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+    }
+
+    /**
+     * Backtracking is required if the green edges in the previous slices are not connected directly or indirectly to the given edge
+     *
+     * @param currentSliceIndex index of the slice from which the edge starts
+     * @param graphEdge         edge of which the predecessors need to verified
+     * @return
+     */
+    public boolean isBacktrackingRequired(int currentSliceIndex, GraphEdge<T, L> graphEdge) {
```
Comment (stephencurran) [inline decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: Does this method need to be `public`?  Is the client going to need to call it?  I don't see any uses of it outside of the class itself.

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+    private boolean isBacktrackingRequired() {
+        return isBacktrackingRequired(currentSourceSliceIndex, graphEdge);
+    }
+
+
+    /**
+     * Forwardtracking is required if there are green successors to the current edge
+     */
+    private boolean isForwardTrackingRequired() {
+        return (graphEdge.getDestinationNode().getOutgoingEdge()
```
Comment (stephencurran) [inline decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: No need for `()` around expression.

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+                .stream()
+                .filter(outgoingEdge -> outgoingEdge.getEdgeColor().equals(EntityColor.GREEN)).count() != 0);
+    }
+
+
+    /**
+     * if no backtracking or forwardtracking is preformed, then next best edge is is estimated based on start or end node of the last best edge
```
Comment (stephencurran) [inline decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: then **the** next best edge **is** estimated

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+            String edgeid = getEdgeId(sourceSliceIndex, sourceNodeIndex, destinationSliceIndex, destinationNodeIndex);
+            graphEdge = edges.get(edgeid);
+            return graphEdge;
+        }
+        return null;
+    }
+
+    /**
+     * Backtrack the graph to identify where the connection is broken and return an unmarked edge with the end node same as the
+     * start node of the last tracked edge. if all edges are already marked then it will try to find an edge from the source slice.
```
Comment (stephencurran) [inline decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: **If** all edges...

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java (hunk 11 above)
Comment (stephencurran) [inline decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: Better to be consistent in terminology.  Sometimes it is "backtracking" and sometimes "backwardtracking".

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java [hunk 13]
```diff
@@ -0,0 +1,305 @@
...
+        if (edge == null) {
+            return getNextEdgeFromNormalFlow();
+        }
+
+        return edge;
+    }
+
+
+    /**
+     * if there are any green successor to the last edge then it will try to find the last green edge in the successor links
```
Comment (stephencurran) [inline decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: **If** there are any green successors

Comment (stephencurran) [reply decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: I find it really hard to understand this really long sentence.  Can you break it into smaller sentences?

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java [hunk 14]
```diff
@@ -0,0 +1,305 @@
...
+            return getNextEdgeFromNormalFlow();
+        }
+
+        return edge;
+    }
+
+    /**
+     * retrieve the next best edge based on the users opinion of the last best edge
+     */
+    private GraphEdge<T, L> edgeFormGraph() {
```
Comment (stephencurran) [inline decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: Should this be called `edgeFromGraph`?

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java [hunk 15]
```diff
@@ -0,0 +1,305 @@
...
+    public boolean isComplete() {
+        return (currentSourceSliceIndex == slices.size() - 1)
+                && lastSuccessfulEdge != null
+                && lastSuccessfulEdge.getEdgeColor().equals(EntityColor.GREEN)
+                && !isBacktrackingRequired(currentSourceSliceIndex - 1, lastSuccessfulEdge);
+    }
+
+
+    /**
+     * returns the next best edge in the graph, and the edge if such an edge actually exist it can be of any color
```
Comment (stephencurran) [inline decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: **Returns**

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java [hunk 16]
```diff
@@ -0,0 +1,305 @@
...
+
+        if (edge == null) {
+            return getNextEdgeFromNormalFlow();
+        }
+
+        return edge;
+    }
+
+    /**
+     * retrieve the next best edge based on the users opinion of the last best edge
```
Comment (stephencurran) [inline decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: **Retrieve**

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java [hunk 17]
```diff
@@ -0,0 +1,305 @@
...
+            generateEdges(sourceSlice, index, destinationSlice, index + 1);
+        }
+    }
+
+    private static String getEdgeId(int sourceSliceIndex, int sourceNodeIndex, int destinationSliceIndex, int destinationNodeIndex) {
+        return sourceSliceIndex + ":" + sourceNodeIndex + ":" + destinationSliceIndex + ":" + destinationNodeIndex;
+    }
+
+    /**
+     * Get the successors of the last Successful edge
```
Comment (stephencurran) [inline decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: **successful**

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java [hunk 18]
```diff
@@ -0,0 +1,305 @@
...
+        List<List<L>> traceSummary = new ArrayList<>();
+        for (GraphEdge<T, L> edge : edges) {
+            traceSummary.add(edge.getValue());
+        }
+        return traceSummary;
+    }
+
+
+    /**
+     * generate edges between the given slices
```
Comment (stephencurran) [inline decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: **Generate**

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java (hunk 2 above)
Comment (stephencurran) [reply decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: Thanks!

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java (hunk 4 above)
Comment (BonyBabu) [reply decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: 31890348b9aef835f9f9065ddd8625de1cfab0ea

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java (hunk 5 above)
Comment (BonyBabu) [reply decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: 31890348b9aef835f9f9065ddd8625de1cfab0ea

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java (hunk 6 above)
Comment (BonyBabu) [reply decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: 31890348b9aef835f9f9065ddd8625de1cfab0ea

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java (hunk 7 above)
Comment (BonyBabu) [reply decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: 31890348b9aef835f9f9065ddd8625de1cfab0ea

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java (hunk 8 above)
Comment (BonyBabu) [reply decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: 31890348b9aef835f9f9065ddd8625de1cfab0ea

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java (hunk 14 above)
Comment (BonyBabu) [reply decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: 31890348b9aef835f9f9065ddd8625de1cfab0ea

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java (hunk 15 above)
Comment (BonyBabu) [reply decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: 31890348b9aef835f9f9065ddd8625de1cfab0ea

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java (hunk 16 above)
Comment (BonyBabu) [reply decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: 31890348b9aef835f9f9065ddd8625de1cfab0ea

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java (hunk 17 above)
Comment (BonyBabu) [reply decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: 31890348b9aef835f9f9065ddd8625de1cfab0ea

File: decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java (hunk 18 above)
Comment (BonyBabu) [reply decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java]: 31890348b9aef835f9f9065ddd8625de1cfab0ea

**Pull request closed without merging**